rul.py

```diff
@@ -0,0 +1,385 @@
+from enum import auto
+from itertools import cycle
+from msilib import sequence
+from random import triangular
+import warnings 
+warnings.filterwarnings("ignore")
+import pandas as pd
+import matplotlib.pyplot as plt
+import numpy as np
+import tensorflow as tf
+import time
+from utilities import *
+from sklearn.metrics import mean_squared_error
+import os
+from datetime import datetime
+
+
+from ECLSTM import ECLSTM1D
+from ECLSTM import check_the_config_valid
+from ECLSTM import build_the_model
+
+from parameters import architecture_parameters
+
+from tensorflow.keras.models import Sequential
+from tensorflow.keras.layers import Dense
+from tensorflow.keras.layers import Flatten
+from tensorflow.keras.layers import TimeDistributed
+from tensorflow.keras.callbacks import TensorBoard, ModelCheckpoint, EarlyStopping, ReduceLROnPlateau
+from typing import Tuple
+
+class RemainingUsefulLife: 
+        
+    def __init__(self,
+                 train_df, 
+                 test_df, 
+                 test_rul_per_rtf_id = None,
+                 train_rul_per_rtf_id = None,
+                 max_life: int = 120, 
+                 sequence_length: int = 5, 
+                 window_size: int = 16, rtf_id = 'rtf_id', cycle_column_name = 'cycle', data_id = 'kaggel_plant', epochs: int = 30, path_to_trained_model: str = None) -> None:
+
+
+        self.__train_df = train_df.copy()
+        self.__test_df = test_df.copy()
+        self.__test_rul_per_rtf_id = test_rul_per_rtf_id 
+        self.__train_rul_per_rtf_id = train_rul_per_rtf_id
+        self.__max_life = max_life
+        self.__sequence_length = sequence_length
+        self.__window_size = window_size
+        self.__rtf_id = rtf_id
+        self.__cycle_column_name = cycle_column_name
+        self.__data_id = data_id
+        self.__epochs = epochs
+        self.__log_dir = path_to_trained_model # one can provide the path to a pretrained model here, if the path is provided, instead of training a new model, the pretrained model is taken to run evaluation
+
+    # storing the string in a variable, since it is often used in the code
+    rul_pw = "RUL_pw"
+
+    # function to calculate the piecewise rul for both training and test data
+    def compute_piecewise_linear_rul(self)->Tuple[pd.DataFrame,pd.DataFrame]:
+        """
+        Computes piecewise linear RUL for the dataframes according to the given max_life paramter 
+
+        Returns:
+            Tuple[pd.DataFrame,pd.DataFrame]: returns training dataset and test dataset with piecewise rul respectively
+        """
+        id = self.__rtf_id
+        datasets = [self.__train_df, self.__test_df]
+        for data in datasets: # since the process for rul calculation is the same, looping through train and test data to calculate pw_rul for each of them 
+            rul = [] 
+
+            # replacing the rtf_ids
```

Comment:
maxLife is input (120)
maxCycleProRTF = data.groupBy("rtf_id", max("cycle") )
data.loc["mxCycle"]  = maxCycleProRTF.???()
data.loc["rul"] = MaxCycle - cycle 

data.loc["pw_rul"] = min(data["rul"], MaxLife)